app/controllers/image_controller.rb

```diff
@@ -300,37 +329,50 @@ def cast_vote
   #   @licenses     (options for license select menu)
   # Redirects to observations/show.
   def add_image
-    pass_query_params
-    @observation = find_or_goto_index(Observation, params[:id].to_s)
-    return unless @observation
-
-    if !check_permission!(@observation)
-      redirect_with_query(controller: :observations,
-                          action: :show, id: @observation.id)
-    elsif request.method != "POST"
-      @image = Image.new
-      @image.license = @user.license
-      @image.copyright_holder = @user.legal_name
-      @image.user = @user
-      # Set the default date to the date of the observation
-      # Don't know how to correctly test this.
-      @image.when = @observation.when
-      @licenses = License.current_names_and_ids(@image.license)
-      init_project_vars_for_add_or_edit(@observation)
-    elsif params[:upload].blank?
+    return unless observation_exists
+
+    check_observation_permission
+
+    return new_image if request.method != "POST"
+
+    create_image
+  end
+
+  private
+
+  def new_image
+    @image = Image.new
+    @image.license = @user.license
+    @image.copyright_holder = @user.legal_name
+    @image.user = @user
+    # Set the default date to the date of the observation
+    # Don't know how to correctly test this.
+    @image.when = @observation.when
+    @licenses = License.current_names_and_ids(@image.license)
+    init_project_vars_for_add_or_edit(@observation)
+  end
+
+  def create_image
+    if params[:upload].blank?
       flash_warning(:runtime_no_changes.t)
-      redirect_with_query(controller: :observations,
-                          action: :show, id: @observation.id)
     else
       args = params[:image]
       i = 1
       while i < 5 || params[:upload]["image#{i}"].present?
         process_image(args, params[:upload]["image#{i}"])
         i += 1
       end
-      redirect_with_query(controller: :observations,
-                          action: :show, id: @observation.id)
     end
+    redirect_with_query(permanent_observation_path(id: @observation.id))
+  end
+
+  def observation_exists
```

Comment:
Ditto.

Reply:
This method is only called once, so consider leaving things as they were?
```ruby
  @observation = ...
  return unless @observation
```
It's transparent, and the class is shorter.

And maybe for the other methods that @pellaea is commenting on.

Reply:
I broke it out because checks like this are repeated between new/create and edit/update actions, in CRUD. But i agree, if i'm assigning with the caller, it's better how it was

But here's a question: do they need to be repeated ? Does the POST action need to check/set the ivar again? I've been doing that, but never sure if necessary

Reply:
I assume POST set it just in case there was an error and it needed to re-serve the form?